src/lib/transition_frontier/full_frontier/full_frontier.ml

```diff
@@ -0,0 +1,566 @@
+            (Breadcrumb.staged_ledger new_root_node.breadcrumb)
+            m0
+        in
+        Ledger.remove_and_reparent_exn m1 m1 ;
+        if Breadcrumb.just_emitted_a_proof new_root_node.breadcrumb then (
+          let s = t.root_ledger in
+          let mt = Ledger.Maskable.register_mask s (Ledger.Mask.create ()) in
+          Non_empty_list.iter
+            (Option.value_exn
+               (Staged_ledger.proof_txns
+                  (Breadcrumb.staged_ledger new_root_node.breadcrumb)))
+            ~f:(fun txn ->
+              ignore (Or_error.ok_exn (Ledger.apply_transaction mt txn)) ) ;
+          Ledger.commit mt ;
+          ignore (Ledger.Maskable.unregister_mask_exn s mt) ) ;
+        new_staged_ledger
+      in
+      let new_root_breadcrumb =
+        Breadcrumb.create
+          (Breadcrumb.validated_transition new_root_node.breadcrumb)
+          new_staged_ledger
+      in
+      let new_root_node =
+        {new_root_node with breadcrumb= new_root_breadcrumb}
+      in
+      Hashtbl.set t.table ~key:new_root_hash ~data:new_root_node ;
+      t.root <- new_root_hash ;
+      Consensus.Hooks.frontier_root_transition
+        (Breadcrumb.consensus_state old_root_node.breadcrumb)
+        (Breadcrumb.consensus_state new_root_node.breadcrumb)
+        ~local_state:t.consensus_local_state ~snarked_ledger:t.root_ledger ;
+      (Breadcrumb.state_hash old_root_node.breadcrumb, Some new_root_hash)
+  (* These are invalid inhabitants for the type signature of this function,
+   * but the OCaml compiler isn't smart enough to realize this. *)
+  | Root_transitioned {garbage= Lite _; _} ->
+      failwith "impossible"
+  | New_node (Lite _) ->
+      failwith "impossible"
+
+let update_metrics_with_diff (type mutant) t
+    (diff : (Diff.full, mutant) Diff.t) : unit =
+  match diff with
+  | New_node _ ->
+      Coda_metrics.(Gauge.inc_one Transition_frontier.active_breadcrumbs) ;
+      Coda_metrics.(Counter.inc_one Transition_frontier.total_breadcrumbs)
+  | Root_transitioned {garbage= Full garbage_breadcrumbs; _} ->
+      let new_root_breadcrumb = root t in
+      let consensus_state = Breadcrumb.consensus_state new_root_breadcrumb in
+      let blockchain_length =
+        Int.to_float
+          ( Consensus.Data.Consensus_state.blockchain_length consensus_state
+          |> Coda_numbers.Length.to_int )
+      in
+      let global_slot =
+        Int.to_float
+          (Consensus.Data.Consensus_state.global_slot consensus_state)
+      in
+      Coda_metrics.(
+        let num_breadcrumbs_removed =
+          Int.to_float (1 + List.length garbage_breadcrumbs)
+        in
+        let num_finalized_staged_txns =
+          Int.to_float
+            (List.length (Breadcrumb.user_commands new_root_breadcrumb))
+        in
+        (* TODO: this metric collection is super inefficient right now *)
+        let root_snarked_ledger_accounts =
+          Ledger.Any_ledger.M.to_list t.root_ledger
```

Comment:
this will probably hurt us when we start testing the larger account-ledger

Reply:
Agreed. There is an open issue for this. This is not new code, this is from develop, so I'm not fixing in this PR.

Reply:
This is removed in develop, so I will be removing this when I handle the merge conflicts.